cfgrammar/src/lib/yacc/parser.rs

```diff
@@ -326,6 +335,38 @@ impl YaccParser {
         }
     }
 
+    fn parse_action(&mut self, i: usize) -> YaccResult<(usize, String)> {
+        let mut j = i;
+        let mut c = 0; // Count braces
+        while j < self.src.len() {
+            let ch = self.src[j..].chars().next().unwrap();
+            c += match ch {
+                '{' => 1,
+                '}' => -1,
+                _ => 0
+            };
+            if c == 0 {
+                break;
+            }
+            j += ch.len_utf8();
+        }
+        if c > 0 {
+            Err(self.mk_error(YaccParserErrorKind::IncompleteAction, j))
+        } else {
+            let s = self.src[i + 1..j - 1].trim().to_string();
+            Ok((j + 1, s))
+        }
+    }
+
+    fn parse_programs(&mut self, i: usize) -> YaccResult<usize> {
```

Comment:
I suggest that we move the `match self.lookahead_is("%%")` from the function that calls `parse_programs` into here. Then we can do something similar to:

```rust
    fn parse_rules(&mut self, mut i: usize) -> YaccResult<usize> {
        // self.parse_declarations should have left the input at '%%'
        match self.lookahead_is("%%", i) {
            Some(j) => i = j,
            None => panic!("Internal error.")
        }
```

Although the code is a bit silly and should probably be:

```rust
    fn parse_rules(&mut self, mut i: usize) -> YaccResult<usize> {
        // self.parse_declarations should have left the input at '%%'
        i = self.lookahead_is("%%").unwrap();
```

Reply:
But then we get a "TypeError" instead of "Internal error" if `self.lookahead_is` returns `None`. Is that ok?

Reply:
Actually, we get an unwrap error.

Reply:
I simplified `parse_rules` and similarly `parse_programs` with the above. Though I can't move the `match self.lookahead_is("%%")` from the function that calls `parse_programs`, since programs are optional and I only want to parse them if there is `%%`. Though, I can do that check within `parse_programs` and make it return `i` if there is no `%%`.

Reply:
OK with me.